commit fe25f99a663027f20aac1a2c719f8af13cce7093
Author:     Dan Davison <dandavison7@gmail.com>
AuthorDate: Sun Jun 30 23:57:10 2019 -0400
Commit:     Dan Davison <dandavison7@gmail.com>
CommitDate: Sun Jun 30 23:57:10 2019 -0400

    By default do not apply syntax highlighting to removed lines

diff --git a/src/main.rs b/src/main.rs
index 311076b..4493f7f 100644
--- a/src/main.rs
+++ b/src/main.rs
@@ -50,6 +50,11 @@ struct Opt {
     /// The syntax highlighting theme to use.
     theme: Option<String>,
 
+    #[structopt(long = "highlight-removed")]
+    /// Apply syntax highlighting to removed lines. The default is to
+    /// apply syntax highlighting to unchanged and new lines only.
+    highlight_removed: bool,
+
     /// The width (in characters) of the diff highlighting. By
     /// default, the highlighting extends to the last character on
     /// each line. By default, the width is equal to the current
@@ -156,6 +161,7 @@ fn process_command_line_arguments<'a>(
         opt.light,
         &opt.plus_color,
         &opt.minus_color,
+        opt.highlight_removed,
         opt.width,
     )
 }
diff --git a/src/paint.rs b/src/paint.rs
index 01e7f05..0b88ed1 100644
--- a/src/paint.rs
+++ b/src/paint.rs
@@ -51,6 +51,7 @@ pub struct Config<'a> {
     minus_color: Color,
     pub syntax_set: &'a SyntaxSet,
     width: Option<usize>,
+    highlight_removed: bool,
     pub pager: &'a str,
 }
 
@@ -61,6 +62,7 @@ pub fn get_config<'a>(
     user_requests_theme_for_light_terminal_background: bool,
     plus_color_str: &Option<String>,
     minus_color_str: &Option<String>,
+    highlight_removed: bool,
     width: Option<usize>,
 ) -> Config<'a> {
 
@@ -95,6 +97,7 @@ pub fn get_config<'a>(
             DARK_THEME_MINUS_COLOR
         }),
         width: width.or_else(|| Some((Term::stdout().size().1 - 1) as usize)),
+        highlight_removed: highlight_removed,
         syntax_set: &syntax_set,
         pager: "less",
     }
@@ -109,6 +112,7 @@ pub fn paint_line(mut line: String, syntax: &SyntaxReference, config: &Config, b
         Some('-') => Some(config.minus_color),
         _ => None,
     };
+    let apply_syntax_highlighting = first_char != Some('-') || config.highlight_removed;
     if first_char == Some('+') || first_char == Some('-') {
         line = line[1..].to_string();
         buf.push_str(" ");
@@ -122,17 +126,33 @@ pub fn paint_line(mut line: String, syntax: &SyntaxReference, config: &Config, b
         _ => (),
     }
     let ranges: Vec<(Style, &str)> = highlighter.highlight(&line, &config.syntax_set);
-    paint_ranges(&ranges[..], background_color, buf);
+    paint_ranges(
+        &ranges[..],
+        background_color,
+        apply_syntax_highlighting,
+        buf,
+    );
 }
 
 /// Based on as_24_bit_terminal_escaped from syntect
 fn paint_ranges(
     foreground_style_ranges: &[(Style, &str)],
     background_color: Option<Color>,
+    apply_syntax_highlighting: bool,
     buf: &mut String,
 ) -> () {
     for &(ref style, text) in foreground_style_ranges.iter() {
-        paint(text, Some(style.foreground), background_color, false, buf);
+        paint(
+            text,
+            if apply_syntax_highlighting {
+                Some(style.foreground)
+            } else {
+                None
+            },
+            background_color,
+            false,
+            buf,
+        );
     }
     buf.push_str("\x1b[0m");
 }
